File: src/main/kotlin/jp/trap/conqest/game/Wallet.java
package jp.trap.conqest.game;

import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;
import org.bukkit.scoreboard.*;

public class Wallet {
    public static void setupScoreboard(Player player){
        ScoreboardManager manager = Bukkit.getScoreboardManager();
        Scoreboard scoreboard = manager.getNewScoreboard();

        Objective objective = scoreboard.registerNewObjective("wallet", "dummy", ChatColor.AQUA + "wallet");
        objective.setDisplaySlot(DisplaySlot.SIDEBAR);

        Score coin = objective.getScore(ChatColor.YELLOW + "Coin");
        coin.setScore(0);
    }

    public static void useCoin(Player player, int use){
        Scoreboard scoreboard = player.getScoreboard();
        Objective objective = scoreboard.getObjective("wallet");
        if(objective == null)return;
        Score score = objective.getScore(ChatColor.YELLOW + "Coin");
        int currentCoin = score.getScore();
        if(currentCoin < use)return;
        score.setScore(currentCoin - use);
    }

    public static void getCoin(Player player, int get){
        Scoreboard scoreboard = player.getScoreboard();
        Objective objective = scoreboard.getObjective("wallet");
        if(objective == null)return;
        Score score = objective.getScore(ChatColor.YELLOW + "Coin");
        int currentCoin = score.getScore();
        if(currentCoin < get)return;
        score.setScore(currentCoin + get);
    }
}

File: src/main/kotlin/jp/trap/conqest/game/Wallet.kt
package jp.trap.conqest.game

import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.scoreboard.DisplaySlot
import org.bukkit.scoreboard.Scoreboard
import org.bukkit.scoreboard.ScoreboardManager
import java.util.UUID

object Wallet : Listener {
    private val OBJECTIVE_NAME = ChatColor.YELLOW.toString() + "Coin"
    private val scoreboardMap = mutableMapOf<UUID, Int>()

    @EventHandler
    fun onPlayerJoin(event: PlayerJoinEvent) {
        val player = event.player
        val uuid = player.uniqueId
        if (scoreboardMap.containsKey(uuid)) {
            setupScoreboard(player, scoreboardMap[uuid]!!)
            scoreboardMap.remove(uuid)
        }
    }

    @EventHandler
    fun onPlayerQuit(event: PlayerQuitEvent) {
        scoreboardMap[event.player.uniqueId] = event.player.scoreboard.getObjective("wallet")?.getScore(OBJECTIVE_NAME)!!.score
    }

    fun setupScoreboard(player: Player, startCoin: Int) {
        val manager = Bukkit.getScoreboardManager()
        val scoreboard = manager.newScoreboard

        val objective = scoreboard.registerNewObjective("wallet", "dummy", ChatColor.AQUA.toString() + "wallet")
        objective.displaySlot = DisplaySlot.SIDEBAR

        val coin = objective.getScore(OBJECTIVE_NAME)
        coin.score = startCoin

        player.scoreboard = scoreboard
    }

    fun removeScoreboard(player: Player) {
        val scoreboard = player.scoreboard
        scoreboard.getObjective("wallet")?.unregister()
        scoreboardMap.remove(player.uniqueId)
    }

    fun pay(player: Player, use: Int): Boolean {
        val scoreboard = player.scoreboard
        val objective = scoreboard.getObjective("wallet") ?: return false
        val score = objective.getScore(OBJECTIVE_NAME)
        val currentCoin = score.score
        if (currentCoin < use) {
            player.sendMessage("コインが足りません")
            return false
        }
        score.score = currentCoin - use
        return true
    }

    fun earn(player: Player, get: Int) {
        val scoreboard = player.scoreboard
        val objective = scoreboard.getObjective("wallet") ?: return
        val score = objective.getScore(OBJECTIVE_NAME)
        val currentCoin = score.score
        score.score = currentCoin + get
    }
}
